fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/AvroDimensionRowParser.java

```diff
@@ -0,0 +1,185 @@
+// Copyright 2016 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.data.dimension;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Parses an AVRO file into Dimension Rows.
+ */
+public class AvroDimensionRowParser {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AvroDimensionRowParser.class);
+    private static ObjectMapper MAPPER = null;
+
+    private DimensionFieldMapper dimensionFieldMapper;
+    private Map<String, String> convertedDimensionFields;
+
+    /**
+     * Constructs an AvroDimensionRowParser object based on the DimensionFieldMapper object.
+     *
+     * @param dimensionFieldMapper Object that defines the dimension field name transformations
+     * @param objectMapper Object that is used to construct mapper objects
+     */
+    public AvroDimensionRowParser(DimensionFieldMapper dimensionFieldMapper, ObjectMapper objectMapper) {
+        this.dimensionFieldMapper = dimensionFieldMapper;
+        MAPPER = objectMapper;
+    }
+
+    /**
+     * Validates the schema of the given AVRO file with the Dimension schema configured by the user.
+     *
+     * @param dimension The dimension object used to configure the dimension
+     * @param avroSchemaPath The path of the AVRO schema file (.avsc)
+     *
+     * @return true if the schema is valid, false otherwise
+     *
+     * @throws NullPointerException thrown if JSON object `fields` is not present
+     * @throws IOException thrown if there is error parsing the avro files
+     */
+    public boolean isSchemaValid(Dimension dimension, String avroSchemaPath) throws NullPointerException, IOException {
+
+        // Convert the AVRO schema file into JsonNode Object
+        JsonNode jsonAvroSchema = convertFileToJsonNode(avroSchemaPath);
+
+        // If the JSON schema file is corrupted/ empty, then return the schema is invalid
+        if (jsonAvroSchema == null) {
+            return false;
```

Comment:
I'm generally a big fan of `Optional::orElseThrow`. In this case, however, I don't see how we could get a `null` back? It looks like (based on tracing the code a little) we'll either get back a `NullNode` (which is not `null`), or we'll get back some other sort of node. The only other cases I can see look like Exception flows, where an exception will be thrown instead of returning anything. So, based on that, do we even need this check at all?